### ISSUE TITLE
twisted.internet.test.test_gireactor.PygtkCompatibilityTests.test_noCompatibilityLayer is broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8144]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8144|
|Type|defect|
|Created|2015-12-15 09:58:36Z|
|Branch|https://github.com/twisted/twisted/tree/pygtk-compat-8144|

```
[ERROR]
Traceback (most recent call last):
  File "/buildslave/ubuntu15.10-py2.7/Twisted/twisted/internet/test/test_gireactor.py", line 218, in test_noCompatibilityLayer
    env={"PYTHONPATH": ":".join(os.path)})
exceptions.TypeError: can only join an iterable

twisted.internet.test.test_gireactor.PygtkCompatibilityTests.test_noCompatibilityLayer
```

We probably want to fix this, since if you have gireactor and no pygtk, it breaks!

<details><summary>Searchable metadata</summary>

```
trac-id__8144 8144
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__None None
branch__branches_pygtk_compat_8144 branches/pygtk-compat-8144
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1450173516534758 1450173516534758
changetime__1450440434740221 1450440434740221
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [46505]) Branching to pygtk-compat-8144.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Okay, this fixes the gireactor bugs -- it passes now (see the 15.10 Py2.7 builder).

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks for the fix, hawkowl.

I am concerned about using `catch_warnings` in this way, because we catch _all_ the warnings and then silently discard the ones we're not interested in.  It's already bad enough that we don't carefully keep our warnings at zero; if we develop the habit of doing this, our test suite will appear to run warnings-clean and then blow up when deprecated functionality is removed from a dependency.  Reading through https://buildbot.twistedmatrix.com/builders/ubuntu15.10-py2.7/builds/317/steps/gi/logs/warnings it does appear as though those warnings are quashed from that function.

Please use a mechanism that will re-emit warnings.  I think the right thing here is `flushWarnings`; both `assertWarns` and `catch_warnings(record=True)` appear to be overzealous and mostly ignore their filtering arguments, whereas `flushWarnings` really does only catch warnings raised directly by the specified function.  If this is not precise enough we might need to create a new warnings filtering thing :-\.

This needs addressing before landing because the responsiveness of our test suite to warnings is an important property to preserve.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl removed owner|
|-|-|

Hi Glyph!

Thanks for the review.

I've altered the branch; it should work fine now :)

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @hawkowl|
|-|-|

lks gd pls lnd

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46523]) Merge pygtk-compat-8144: Fix pygtk compat tests and make gireactor a supported platform.

Author: hawkowl
Reviewer: glyph
Fixes: [#8144](https://github.com/twisted/twisted/issues/8144)
